src/reconstruction.jl

```diff
@@ -0,0 +1,212 @@
+The most recent values in the reconstruction are located at the end of each vector.
+
+A simpler constructor for convenience
+```julia
+	SpatioTemporalEmbedding(s, D, τ, B, k, ::Type{<:AbstractBoundaryCondition})
+```
+Takes as arguments the spatial timeseries `s` and reconstructs
+`B` spatial shells separated by `k` points around each point
+and repeats this for `D` past timesteps separated by `τ` each.
+"""
+struct SpatioTemporalEmbedding{T,Φ,BC,X} <: AbstractSpatialEmbedding{T,Φ,BC,X}
+  	τ::Vector{Int}
+	β::Vector{CartesianIndex{Φ}}
+	inner::Region{Φ}  #inner field far from boundary
+	whole::Region{Φ}
+
+	function SpatioTemporalEmbedding{T,Φ,BC,X}(τ,β,fsize) where {T,Φ,BC,X}
+		inner = inner_region(β, fsize)
+		whole = Region((ones(Int,Φ)...,), fsize)
+		return new{T,Φ,BC,X}(τ,β,inner,whole)
+	end
+end
+const STE = SpatioTemporalEmbedding
+
+function SpatioTemporalEmbedding(
+		s::AbstractArray{<:AbstractArray{T,Φ}},
+		D, τ, B, k, ::Type{BC}
+		) where {T,Φ, BC<:AbstractBoundaryCondition}
+	@assert issorted(τ) "Delays need to be sorted in ascending order"
+	X = (D+1)*(2B+1)^Φ
+	τs = Vector{Int}(undef,X)
+	βs = Vector{CartesianIndex{Φ}}(undef,X)
+	n = 1
+	for d=0:D, α = Iterators.product([-B*k:k:B*k for φ=1:Φ]...)
+		τs[n] = d*τ
+		βs[n] = CartesianIndex(α)
+		n +=1
+	end
+	return SpatioTemporalEmbedding{T,Φ,BC,X}(τs, βs, size(s[1]))
+end
+
+
+
+#This function is not safe. If you call it directly with bad params - can fail
+function (r::SpatioTemporalEmbedding{T,Φ,ConstantBoundary{C},X})(rvec,s,t,α) where {T,Φ,C,X}
+	if α in r.inner
+		@inbounds for n=1:X
+			rvec[n] = s[ t + r.τ[n] ][ α + r.β[n] ]
+		end
+	else
+		@inbounds for n=1:X
+			rvec[n] = α + r.β[n] in r.whole ? s[ t+r.τ[n] ][ α+r.β[n] ] : C
+		end
+	end
+	return nothing
+end
+
+function (r::SpatioTemporalEmbedding{T,Φ,PeriodicBoundary,X})(rvec,s,t,α) where {T,Φ,X}
+	if α in r.inner
+		@inbounds for n=1:X
+			rvec[n] = s[ t + r.τ[n] ][ α + r.β[n] ]
+		end
+	else
+		@inbounds for n=1:X
+			rvec[n] = s[ t + r.τ[n] ][ project_inside(α + r.β[n], r.whole) ]
+		end
+	end
+	return nothing
+end
+
+
+
+get_num_pt(em::SpatioTemporalEmbedding) = prod(em.whole.maxi)
+get_τmax(em::SpatioTemporalEmbedding{T,Φ,BC,X}) where {T,Φ,BC,X} = em.τ[X]
+outdim(em::SpatioTemporalEmbedding{T,Φ,BC,X}) where {T,Φ,BC,X} = X
+
+
+
+
+function Base.summary(::IO, ::SpatioTemporalEmbedding{T,Φ,BC, X}) where {T,Φ,BC,X}
+	println("$(Φ)D Spatio-Temporal Delay Embedding with $X Entries")
+end
+
+
+"""
+	reconstruct(s::AbstractArray{<:AbstractArray{T,Φ}}, em)
+Reconstruct the spatial timeseries `s` represented by a `Vector` of `AbstractArray`
+states using the embedding struct `em` of type [`AbstractSpatialEmbedding`](@ref).
+
+Returns the reconstruction in the form of a [`Dataset`](@ref) where each row is a
+reconstructed state and they are ordered first through linear indexing into each state
+and then incrementing in time.
+"""
+function reconstruct(s::AbstractArray{<:AbstractArray{T,Φ}},
```

Comment:
Wait is this still going on? `Dataset` has everything as `SVector`s, which is super inneficient for large dimensions.

Reply:
Large dimensions are super inefficient for nearest neighbor searches as well.
So we typically won't be calling `reconstruct` for dimensions > 50.
If I use a matrix, `KDTree` coverts it into vector or svectors internally anyway.

Reply:
BTW this is the reason for the slightly funky _inplace_ embedding functor
like `em(recv, s, t, α)`. 
Doing this out of place by returning an `SVector` becomes slow for large dimensions.
Particularly so for the `PCAEmbedding` as I have to multiply it with a Matrix.